### DISCUSSION TITLE
Create a separate sheet for each category in a column

### DISCUSSION THREAD

**Discussion by ghost** [saulpw/visidata#1151] [Q&A] [answered]
Sometimes I have a single sheet with a few different categories in a certain column, but I want multiple sheets, one for each category.

I can do this by hitting `Shift-F`, then pressing `Enter` on the first category, then returning to the frequency sheet, and pressing `Enter` on the second category, and so on...

Is there a quicker way to do this?

**Comment by saulpw** [comment, marked as answer]
Everything is possible with the magic of Python! :)

I think something like this should work:

```
@FreqTableSheet.api
def freq_subsheets(sheet, rows):
    for r in rows:
        vs = copy(sheet.source)
        vs.name += "_" + valueNames(r.discrete_keys, r.numeric_key)
        vs.rows = copy(r.sourcerows)
        yield vs

FreqTableSheet.addCommand('g'+ENTER, 'open-rows', 'for vs in freq_subsheets(selectedRows): vd.push(vs)')
```

(It's important to put the `push`, `selectedRows` in the execstr itself (along with any `cursor*` or `input()`), as various things like the cmdlog and menu system look into the execstr for their functionality.)

[edited to include @davidwales improvements below]

**Reply by ghost** [reply]
That works perfectly! Thank you!

I borrowed a line from [freqtbl.py](https://github.com/saulpw/visidata/blob/dc8cef4980bdef07fd3640be8c92b9e3dc8be9dc/visidata/freqtbl.py#L66-L73) to include the name of the category as would normally occur when pressing `Enter` in a `FrequencyTableSheet`:

``` python
from visidata.freqtbl import valueNames
@FreqTableSheet.api
def freq_subsheets(sheet, rows):
    for r in rows:
        vs = copy(sheet.source)
        vs.name += "_" + valueNames(r.discrete_keys, r.numeric_key)
        vs.rows = r.sourcerows
        yield vs

FreqTableSheet.addCommand('z'+ENTER, 'open-rows', 'for vs in freq_subsheets(selectedRows): vd.push(vs)')
```

I noticed in `freqtable.py` that the `vs.rows = r.sourcerows` uses `copy(r.sourcerows)`. Is that something I should include here too?

**Reply by saulpw** [reply]
Yes, probably should, for good measure.  Otherwise if you do this command multiple times (or other commands which use `sourcerows` directly), then deleting a row on one sheet will cause that row to be deleted from all of the sheets, which is probably not what you want.

**Reply by frosencrantz** [reply]
It would be great if this function could be added to Visidata. I was recently doing something and wanted this exact functionality.  I accomplished that by going back to the freqtable multiple times to create the tables I wanted.

**Reply by frosencrantz** [reply]
Would it make sense to add this as a wishlist request?

**Reply by saulpw** [reply]
If other people want this, we could certainly add this code snippet to the freqtbl basically as-is.  Is `z Enter` the correct binding?

**Reply by ghost** [reply]
The binding's a bit confusing. A similar operation in a DirSheet or an XslxIndexSheet uses `g Enter`. But `g Enter` is already taken in the FreqTableSheet for opening the respective rows from the source sheet.

|sheet               |longname            |keystrokes          |description                                                                                                                                                 |
|--------------------|--------------------|--------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------|
|Canvas              |dive\-cursor        |Enter               |open sheet of source rows contained within canvas cursor                                                                                                    |
|ChoiceSheet         |choose\-rows        |Enter               |                                                                                                                                                            |
|ColumnsSheet        |freq\-row           |Enter               |open a Frequency Table sheet grouped on column referenced in current row                                                                                    |
|CrosswordSheet      |open\-grid          |Enter               |open GridSheet: grid for crossword                                                                                                                          |
|CrosswordsSheet     |open\-clues         |Enter               |open CrosswordSheet: clue answer pair for crossword                                                                                                         |
|DirSheet            |open\-row           |Enter               |open current file as a new sheet                                                                                                                            |
|HelpSheet           |exec\-command       |Enter               |execute command on undersheet                                                                                                                               |
|PandasFreqTableSheet|open\-row           |Enter               |open copy of source sheet with rows that are grouped in current row                                                                                         |
|SheetsSheet         |open\-row           |Enter               |open sheet referenced in current row                                                                                                                        |
|TableSheet          |open\-row           |Enter               |open sheet with copies of rows referenced in current row                                                                                                    |
|TextCanvas          |dive\-cursor        |Enter               |dive into source rows under cursor                                                                                                                          |
|Canvas              |dive\-visible       |gEnter              |open sheet of source rows visible on screen                                                                                                                 |
|ChoiceSheet         |choose\-rows\-selected|gEnter              |                                                                                                                                                            |
|FreqTableSheet      |dive\-rows          |gEnter              |open copy of source sheet with rows that are grouped in selected rows                                                                                       |
|TableSheet          |dive\-selected      |gEnter              |open sheet with copies of rows referenced in selected rows                                                                                                  |
|TextCanvas          |dive\-selected      |gEnter              |dive into selected source rows                                                                                                                              |
|TableSheet          |dive\-selected\-cells|gzEnter             |open sheet with copies of rows referenced in selected rows                                                                                                  |
|DirSheet            |open\-rows          |zEnter              |open selected files as new sheets                                                                                                                           |
|FreqTableSheet      |open\-rows          |zEnter              |\(\.visidatarc command\) open a separate copy of source sheet with rows that are grouped in each selected frequency bin                                     |
|TableSheet          |open\-cell          |zEnter              |open sheet with copies of rows referenced in current cell                                                                                                   |

**Reply by saulpw** [reply]
Maybe `g Enter` should always be `open-rows`, pushing multiple sheets, and here `z Enter` then could be `dive-rows`, pushing one combined sheet.  What do people think about this?  It's an interface change but it would be more consistent.

**Reply by ghost** [reply]
I like it. :)

**Reply by frosencrantz** [reply]
I use the current FreqTableSheet `g Enter` `dive-rows`, and I think I would use it more often then the new command.  I would make use of both.  If you think this would make the bindings more consistent, I will grumble a bit, but that seems reasonable.

**Reply by jsvine** [reply]
FWIW, I'm also a frequent — pun not intended but preserved ;) — user of `g-Enter` on the Frequency Sheet. Though I also, like @frosencrantz, would understand if you tweaked the bindings for consistency's sake.

**Reply by saulpw** [reply]
Another possibility is `go` for `open-rows`; does that make more sense to anyone?

**Reply by ghost** [reply]
It's been said that 'a foolish consistency is the hobgoblin of little minds', so perhaps I have a little mind... But if we did `go` for open rows, would we also change DirSheet and XlsxSheet to use `go` to open multiple sheets?

And if we're using `go` to open multiple sheets, would we then use `o` to open single sheets?

Looking at the help text, `open-row` is `Enter` across a few different sheet types, while `open-rows` is `zEnter`, and `dive-*` is `g?Enter`.
This seems to be consistent at least. 

However, although my help text suggests that `open-rows` is `zEnter` for DirSheets, I think the help text is wrong, because `gEnter` is definitely used here.

If we ensure that `open-rows` in DirSheet and XlsxSheet are changed to use `zEnter`, then we can keep `gEnter` for `dive-*`, and `Enter` for `open-row`, and everything should be consistent? (Or vice versa, swapping `g` and `z`.)

I'm also thinking about how `g` usually means 'make it BIGGER' and `z` usually means 'make it sᴍᴀʟʟᴇʀ', so if we consider that if `Enter` is `open-row`, is it bigger or smaller to `open-rows`, or to `dive-*`?

I don't really mind what the shortcut is, so long as it's consistent! 😄

**Reply by saulpw** [reply]
Okay, I am thinking that `open-row`, `open-rows-multisheet`, and `open-rows` should all be generic commands on the basic TableSheet, and then they'll be consistent by design. We need a better name than `open-rows-multisheet` though.  `open-rows-each`?  any ideas?

**Reply by ghost** [reply]
I'm assuming that `open-rows-multisheet` means 'open each row as a separate sheet, while `open-rows` means 'open a single sheet with all selected rows'?

Perhaps `open-rows-sheets`?

**Reply by saulpw** [reply]
okay, we can add `open-rows-sheets` as a possibility too.  It's a bit of a mouthful though :)
